pandas/_libs/tslib.pyx

```diff
@@ -445,6 +445,8 @@ cpdef array_to_datetime(
     bint utc=False,
     bint require_iso8601=False,
     bint allow_mixed=False,
+    str format="",
```

Comment:
Can you use None as a default rather than `""`? I think can just type as Optional[str] which would be a bit cleaner

Reply:
If None can be interpreted as nullptr, I agree that having None is a cleaner option.

Reply:
This isn't user-facing, it's just used internally

Not sure I understand about None being interpreted as nullptr, but if I pass `None` to `get_c_string_buf_and_size` I get
```
Traceback (most recent call last):
  File "t.py", line 109, in <module>
    print(to_datetime('2012-01-01T10:00', format='%Y-%m-%dT%H:%M', exact=False))
  File "/home/marcogorelli/pandas-dev/pandas/core/tools/datetimes.py", line 1117, in to_datetime
    result = convert_listlike(np.array([arg]), format)[0]
  File "/home/marcogorelli/pandas-dev/pandas/core/tools/datetimes.py", line 441, in _convert_listlike_datetimes
    result, tz_parsed = objects_to_datetime64ns(
  File "/home/marcogorelli/pandas-dev/pandas/core/arrays/datetimes.py", line 2177, in objects_to_datetime64ns
    result, tz_parsed = tslib.array_to_datetime(
  File "pandas/_libs/tslib.pyx", line 440, in pandas._libs.tslib.array_to_datetime
    cpdef array_to_datetime(
  File "pandas/_libs/tslib.pyx", line 706, in pandas._libs.tslib.array_to_datetime
    raise
  File "pandas/_libs/tslib.pyx", line 606, in pandas._libs.tslib.array_to_datetime
    string_to_dts_failed = string_to_dts(
  File "pandas/_libs/tslibs/np_datetime.pyx", line 288, in pandas._libs.tslibs.np_datetime.string_to_dts
    format_buf = get_c_string_buf_and_size(None, &format_length)
  File "pandas/_libs/tslibs/util.pxd", line 221, in pandas._libs.tslibs.util.get_c_string_buf_and_size
    return PyUnicode_AsUTF8AndSize(py_string, length)
TypeError: bad argument type for built-in operation
```

So I think the conversion between `None` and `str` needs to happen at some point internally anyway

If `exact` if False and `format` is `''`, then it should match anything

Reply:
What about?
```python
if format is not None: 
  format_buf = 0
  format_lenght = 0
else:
  format_buf = get_c_string_buf_and_size(...)
...

return parse_iso...
```
The rest of the code should work just fine given that the checks are conditioned on `format_lenght`.

Reply:
As long as `format_len` is `0`, `*format_buf` could be anything, so it is a bit clearer to have it as `NULL`, imho.

Reply:
how though? because if it's `0`, then we get
```
  File "t.py", line 109, in <module>
    print(to_datetime('2012-01-01T10:00', format='%Y-%m-%dT%H:%M', exact=False))
  File "/home/marcogorelli/pandas-dev/pandas/core/tools/datetimes.py", line 1117, in to_datetime
    result = convert_listlike(np.array([arg]), format)[0]
  File "/home/marcogorelli/pandas-dev/pandas/core/tools/datetimes.py", line 441, in _convert_listlike_datetimes
    result, tz_parsed = objects_to_datetime64ns(
  File "/home/marcogorelli/pandas-dev/pandas/core/arrays/datetimes.py", line 2177, in objects_to_datetime64ns
    result, tz_parsed = tslib.array_to_datetime(
  File "pandas/_libs/tslib.pyx", line 440, in pandas._libs.tslib.array_to_datetime
    cpdef array_to_datetime(
  File "pandas/_libs/tslib.pyx", line 706, in pandas._libs.tslib.array_to_datetime
    raise
  File "pandas/_libs/tslib.pyx", line 606, in pandas._libs.tslib.array_to_datetime
    string_to_dts_failed = string_to_dts(
  File "pandas/_libs/tslibs/np_datetime.pyx", line 287, in pandas._libs.tslibs.np_datetime.string_to_dts
    format_buf = get_c_string_buf_and_size(0, &format_length)
TypeError: Expected unicode, got int
```

Reply:
I've tried
```
format_buf = <const char*> 0
```
but then get
```
Segmentation fault
```
(with no further info)

For now I've gone with `format = b''`, and handling `None` inside this function does indeed look cleaner, thanks!